src/common/scram.c

```diff
@@ -0,0 +1,297 @@
+		EVP_MD_CTX_free(md_ctx);
+		return SCRAM_ERROR;
+	}
+
+	if (!EVP_DigestUpdate(md_ctx, input, input_len)) {
+		session->error = g_strdup("Message digest update failed");
+		EVP_MD_CTX_free(md_ctx);
+		return SCRAM_ERROR;
+	}
+
+	if (!EVP_DigestFinal_ex(md_ctx, output, output_len)) {
+		session->error = g_strdup("Message digest finalization failed");
+		EVP_MD_CTX_free(md_ctx);
+		return SCRAM_ERROR;
+	}
+
+	EVP_MD_CTX_free(md_ctx);
+	return SCRAM_IN_PROGRESS;
+}
+
+static int process_client_first(scram_session *session, char **output, size_t *output_len)
+{
+	char nonce[NONCE_LENGTH];
+
+	if (!create_nonce(nonce, NONCE_LENGTH)) {
+		session->error = g_strdup("Could not create client nonce");
+		return SCRAM_ERROR;
+	}
+
+	session->client_nonce_b64 = g_base64_encode((guchar *) nonce, NONCE_LENGTH);
+	*output = g_strdup_printf("n,,n=%s,r=%s", session->username, session->client_nonce_b64);
+	*output_len = strlen(*output);
+	session->client_first_message_bare = g_strdup(*output + 3);
+	session->step++;
+	return SCRAM_IN_PROGRESS;
+}
+
+static int process_server_first(scram_session *session, const char *data, char **output,
+                                size_t *output_len)
+{
+	char **params, *client_final_message_without_proof, *salt, *server_nonce_b64,
+	    *client_proof_b64;
+	unsigned char *client_key, stored_key[EVP_MAX_MD_SIZE], *client_signature, *client_proof;
+	unsigned int i, param_count, salt_len, iteration_count, client_key_len, stored_key_len;
+	size_t client_nonce_len;
+	gint b64_state = 0;
+	guint b64_save = 0;
+
+	params = g_strsplit(data, ",", -1);
+	param_count = g_strv_length(params);
+
+	if (param_count < 3) {
+		session->error = g_strdup_printf("Invalid server-first-message: %s", data);
+		g_strfreev(params);
+		return SCRAM_ERROR;
+	}
+
+	server_nonce_b64 = NULL;
+	salt = NULL;
+	iteration_count = 0;
+
+	for (i = 0; i < param_count; i++) {
+		if (!strncmp(params[i], "r=", 2)) {
+			server_nonce_b64 = g_strdup(params[i] + 2);
+		} else if (!strncmp(params[i], "s=", 2)) {
+			salt = g_strdup(params[i] + 2);
+		} else if (!strncmp(params[i], "i=", 2)) {
+			iteration_count = strtoul(params[i] + 2, NULL, 10);
+		}
+	}
+
+	g_strfreev(params);
+
+	if (server_nonce_b64 == NULL || *server_nonce_b64 == '\0' || salt == NULL ||
+	    *salt == '\0' || iteration_count == 0) {
+		session->error = g_strdup_printf("Invalid server-first-message: %s", data);
+		return SCRAM_ERROR;
+	}
+
+	client_nonce_len = strlen(session->client_nonce_b64);
+
+	// The server can append his nonce to the client's nonce
+	if (strlen(server_nonce_b64) < client_nonce_len ||
+	    strncmp(server_nonce_b64, session->client_nonce_b64, client_nonce_len)) {
+		session->error = g_strdup_printf("Invalid server nonce: %s", server_nonce_b64);
+		return SCRAM_ERROR;
+	}
+
+	salt_len = g_base64_decode_step((gchar *) salt, strlen((char *) salt), (guchar *) salt,
+	                                &b64_state, &b64_save);

```

Comment:
Everywhere I see you use `g_base64_decode_step()` it seems like `g_base64_decode_inplace()` would be more appropriate?